projects/storefrontlib/src/cms-structure/services/feature-modules.service.ts

```diff
@@ -0,0 +1,286 @@
+      'featureModules'
+    );
+
+    this.featureModulesConfig = config.featureModules ?? {};
+
+    for (const [featureName, featureConfig] of Object.entries(
+      this.featureModulesConfig
+    )) {
+      if (featureConfig.cmsComponents?.length) {
+        for (const component of featureConfig.cmsComponents) {
+          this.componentFeatureMap.set(component, featureName);
+        }
+      }
+    }
+  }
+
+  /**
+   * Check if there is feature module configuration that covers specified
+   * component type
+   */
+  hasFeatureFor(componentType: string): boolean {
+    return this.componentFeatureMap.has(componentType);
+  }
+
+  /**
+   * Return full CmsComponent mapping defined in feature module
+   */
+  getCmsMapping(componentType: string): Observable<CmsComponentMapping> {
+    const feature = this.componentFeatureMap.get(componentType);
+
+    return this.resolveFeature(feature).pipe(
+      map(
+        (featureInstance) => featureInstance.componentsMappings[componentType]
+      )
+    );
+  }
+
+  /**
+   * Get all injectors for feature and its dependencies
+   *
+   * As it's a synchronous method, it works only for already resolved features,
+   * returning undefined otherwise
+   */
+  getInjectors(componentType: string): Injector[] | undefined {
+    const feature = this.componentFeatureMap.get(componentType);
+    let injectors;
+
+    // we are returning injectors only for already resolved features
+    this.features
+      .get(feature)
+      ?.subscribe((featureInstance) => {
+        injectors = [
+          // feature module injector
+          featureInstance.moduleRef.injector,
+          // injectors from dependency modules
+          ...featureInstance.depsModules.map(
+            (module) => this.dependencyModules.get(module).injector
+          ),
+        ];
+      })
+      .unsubscribe();
+    return injectors;
+  }
+
+  /**
+   * Resolve feature based on feature name, if feature was not yet resolved
+   *
+   * It will first resolve all module dependencies if defined
+   */
+  private resolveFeature(featureName: string): Observable<FeatureInstance> {
+    return defer(() => {
+      if (!this.features.has(featureName)) {
+        const featureConfig = this.featureModulesConfig[featureName];
+
+        if (!featureConfig?.module) {
+          throw new Error(
+            'No module defined for Feature Module ' + featureName
+          );
+        }
+
+        // resolve dependencies first (if any)
+        const depsResolve = featureConfig.dependencies?.length
+          ? forkJoin(
+              featureConfig.dependencies.map((depModuleFunc) =>
+                this.resolveDependencyModule(depModuleFunc)
+              )
+            )
+          : of(undefined);
+
+        this.features.set(
+          featureName,
+          depsResolve.pipe(
+            switchMap((deps) => this.resolveFeatureModule(featureConfig, deps)),
+            shareReplay()
```

Comment:
We didn't specify the buffer size 1. Are we absolutely sure the upper observable won't emit more than once? I recall problems with `resolveModuleFactory` and `observeOn` which makes me suspicious.

Anyway, I I wouldn't mind passing buffer size 1 to be bullet-proof for future

Reply:
thx!